arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,234 @@
+        auto hashIt = _sstFileNamesToHashes.find(*it);
+        if (hashIt == _sstFileNamesToHashes.end()) {
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          std::string tempPath =
+              basics::FileUtils::buildFilename(_rootPath, *it);
+          LOG_TOPIC("d6c86", DEBUG, arangodb::Logger::ENGINES)
+              << "checkMissingShaFiles:"
+                 " Computing cheacksum for "
+              << tempPath;
+          auto checksumCalc = ChecksumCalculator();
+          if (TRI_ProcessFile(tempPath.c_str(),
+                              [&checksumCalc](char const* buffer, size_t n) {
+                                checksumCalc.update(buffer, n);
+                                return true;
+                              })) {
+            writeShaFile(tempPath, checksumCalc.computeChecksum());
+          }
+        }
+      }
+    }
+  }
+}
+
+rocksdb::Status ChecksumEnv::NewWritableFile(
+    const std::string& fileName, std::unique_ptr<rocksdb::WritableFile>* result,
+    const rocksdb::EnvOptions& options) {
+  std::unique_ptr<rocksdb::WritableFile> writableFile;
+  rocksdb::Status s =
+      rocksdb::EnvWrapper::NewWritableFile(fileName, &writableFile, options);
+  if (!s.ok()) {
+    return s;
+  }
+  result->reset(
+      new ChecksumWritableFile(writableFile.release(), fileName, _helper));
+  return s;
+}
+
+rocksdb::Status ChecksumWritableFile::Close() {
+  if (!_helper->isFileNameSst(_sstFileName)) {
+    return rocksdb::WritableFileWrapper::Close();
+  }
+  TRI_ASSERT(_helper != nullptr);
+  auto checksumCalc = ChecksumCalculator();
+  if (TRI_ProcessFile(_sstFileName.c_str(),
+                      [&checksumCalc](char const* buffer, size_t n) {
+                        checksumCalc.update(buffer, n);
+                        return true;
+                      })) {
+    if (_helper->writeShaFile(_sstFileName, checksumCalc.computeChecksum())) {
+      return rocksdb::Status::OK();
+    }
+  }
+  return rocksdb::Status::Aborted("File writing was unsuccessful");
+}
+
+rocksdb::Status ChecksumEnv::DeleteFile(const std::string& fileName) {
+  if (!_helper->isFileNameSst(fileName) &&
+      fileName.find(".sha") == std::string::npos) {
+    return rocksdb::EnvWrapper::DeleteFile(fileName);
+  }
+  return _helper->DeleteFile(fileName);
+}
+
+rocksdb::Status ChecksumHelper::DeleteFile(const std::string& fileName) {
+  std::string shaFileName;
+  {
+    MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+    if (auto it = _sstFileNamesToHashes.find(TRI_Basename(fileName));
+        it != _sstFileNamesToHashes.end()) {
+      TRI_ASSERT(fileName.size() > 4);
+      shaFileName.append(fileName, 0,
+                         fileName.size() - 4);  // append without .sst
+      TRI_ASSERT(!isFileNameSst(shaFileName));
+      shaFileName += ".sha." + (*it).second + ".hash";
+      _sstFileNamesToHashes.erase(it);
+    }
+    if (!shaFileName.empty()) {
+      auto res = TRI_UnlinkFile(shaFileName.data());
```

Comment:
Maybe we want to use the downward env's `DeleteFile` here rather than messing with the file in the filesystem ourselves?